File: verification/verify_crypto.mjs

import { webcrypto } from 'node:crypto';
import fs from 'node:fs';
import { execSync } from 'node:child_process';

// Polyfill window environment for core.js
global.window = {
    crypto: webcrypto,
    atob: atob,
    btoa: btoa
};

// Import the function to test
// Note: We need to import the file directly.
// Since core.js uses 'export', we can import it.
import { signRequest } from '../kalshi-dashboard/src/utils/core.js';

async function testCrypto() {
    console.log("Generating RSA Private Key (PKCS#1)...");
    try {
        execSync('openssl genrsa -out test_key.pem 2048');
    } catch (e) {
        console.error("OpenSSL failed. Is it installed?");
        process.exit(1);
    }

    const privateKeyPem = fs.readFileSync('test_key.pem', 'utf8');
    console.log("Key generated. Testing signRequest...");

    try {
        const timestamp = Date.now();
        const method = "GET";
        const path = "/trade-api/v2/markets";

        const signature = await signRequest(privateKeyPem, method, path, timestamp);

        console.log("Signature generated successfully!");
        console.log("Signature length:", signature.length);
        console.log("Signature (Base64):", signature.substring(0, 50) + "...");

        // Basic validation: signature should be non-empty string
        if (!signature || typeof signature !== 'string' || signature.length < 100) {
            throw new Error("Invalid signature generated");
        }

        // Test PKCS#8 Key (convert existing key)
        console.log("\nConverting to PKCS#8 and testing...");
        execSync('openssl pkcs8 -topk8 -inform PEM -outform PEM -nocrypt -in test_key.pem -out test_key_pkcs8.pem');
        const pkcs8Key = fs.readFileSync('test_key_pkcs8.pem', 'utf8');

        const sig8 = await signRequest(pkcs8Key, method, path, timestamp);
        console.log("PKCS#8 Signature generated successfully!");

        // Clean up
        fs.unlinkSync('test_key.pem');
        fs.unlinkSync('test_key_pkcs8.pem');

        console.log("\n✅ VERIFICATION PASSED: Both PKCS#1 and PKCS#8 keys supported via Native Web Crypto.");

    } catch (error) {
        console.error("❌ VERIFICATION FAILED:", error);
        // Clean up
        try { fs.unlinkSync('test_key.pem'); } catch {}
        try { fs.unlinkSync('test_key_pkcs8.pem'); } catch {}
        process.exit(1);
    }
}

testCrypto();

File: verification/verify_crypto.py
import os
import shutil
import subprocess
import time
import sys
from playwright.sync_api import sync_playwright
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import rsa

def generate_key():
    key = rsa.generate_private_key(
        public_exponent=65537,
        key_size=2048,
    )
    pem = key.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption()
    )
    return pem.decode('utf-8')

def setup_test_env():
    test_dir = "verification/crypto_test_env"
    if os.path.exists(test_dir):
        shutil.rmtree(test_dir)
    os.makedirs(test_dir)

    # Copy core.js
    shutil.copy("kalshi-dashboard/src/utils/core.js", f"{test_dir}/core.js")

    # Create index.html
    html_content = """
    <!DOCTYPE html>
    <html>
    <body>
    <script type="module">
        import { signRequest } from './core.js';

        window.runTest = async (pem) => {
            try {
                const signature = await signRequest(pem, "GET", "/test", Date.now());
                console.log("Signature generated:", signature);
                return signature;
            } catch (e) {
                console.error(e);
                return "ERROR: " + e.message;
            }
        };
    </script>
    </body>
    </html>
    """
    with open(f"{test_dir}/index.html", "w") as f:
        f.write(html_content)

    return test_dir

def run_server(test_dir, port=8001):
    # Start python server in background
    proc = subprocess.Popen([sys.executable, "-m", "http.server", str(port)], cwd=test_dir, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    time.sleep(1) # wait for start
    return proc

def verify():
    print("Setting up test environment...")
    test_dir = setup_test_env()

    print("Generating PKCS#8 Key...")
    pem_key = generate_key()

    print("Starting server...")
    server_proc = run_server(test_dir)

    try:
        with sync_playwright() as p:
            print("Launching browser...")
            browser = p.chromium.launch()
            page = browser.new_page()

            print("Navigating to test page...")
            page.goto("http://localhost:8001")

            print("Running signRequest...")
            # We pass the PEM key to the function exposed on window
            result = page.evaluate(f"window.runTest(`{pem_key}`)")

            print(f"Result: {result}")

            if "ERROR" in str(result):
                print("❌ Verification Failed: Error returned from browser")
                sys.exit(1)
            elif result and len(result) > 10:
                print("✅ Verification Passed: Signature generated successfully")
            else:
                print("❌ Verification Failed: Invalid result")
                sys.exit(1)

            browser.close()

    finally:
        server_proc.kill()
        # Clean up
        if os.path.exists(test_dir):
            shutil.rmtree(test_dir)

if __name__ == "__main__":
    verify()
